Reduce use of memcpy in ConnectionUnix.cpp

AlexsanderDamaceno opened the pull request:
#### 9d2429975387f5899f3a05ed4b7bdac401537185
<pre>
Reduce use of memcpy in ConnectionUnix.cpp
<a href="https://bugs.webkit.org/show_bug.cgi?id=290511">https://bugs.webkit.org/show_bug.cgi?id=290511</a>

Reviewed by Chris Dumez.

Reduce use of memcpy in ConnectionUnix, memcpy is unsafe and can lead to memory corruption bugs.

* Source/WebKit/Platform/IPC/unix/ConnectionUnix.cpp:
(IPC::Connection::processMessage):

Canonical link: <a href="https://commits.webkit.org/292897@main">https://commits.webkit.org/292897@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/73b45fc6ab4ad64424e1aba28074e099e859e714

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/97262 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/16887 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/7101 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/102347 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/47790 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/17180 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/25336 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/74110 "Passed tests") | [❌ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/31302 "Found 1 new test failure: workers/worker-to-worker.html (failure)") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/100265 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/12991 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/87963 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/54447 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/12745 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/47232 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/82792 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/5924 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/104368 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/24339 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/17756 "Found 1 new test failure: imported/w3c/web-platform-tests/workers/abrupt-completion.html (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/83154 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/24712 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/84087 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/82563 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/20800 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/27130 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/4793 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/17884 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/24303 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/29555 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/24125 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/27439 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/25699 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebKit/Platform/IPC/unix/ConnectionUnix.cpp:
```diff
@@ -178,7 +178,7 @@ bool Connection::processMessage()
 
     if (attachmentCount) {
-        memcpy(static_cast<void*>(attachmentInfo.data()), messageData, sizeof(AttachmentInfo) * attachmentCount);
+        memcpySpan(asMutableByteSpan(attachmentInfo), m_readBuffer.span().first(attachmentCount));
```
```suggestion
        memcpySpan(asMutableByteSpan(attachmentInfo), m_readBuffer.span().first(sizeof(AttachmentInfo) * attachmentCount));
```

mcatanzaro replied:
But how do we advance `m_readerBuffer` past the `messageInfo`? We need to compensate for the `messageData += sizeof(messageInfo)`.

cdumez replied:
You're right, the code still wouldn't be correct, even after Darin's proposal. We should really use `messageData` and make it a span. This code is still doing pointer arithmetic and is thus very much unsafe still.

AlexsanderDamaceno replied:
@mcatanzaro  thank you for note this, had not noticed this offset displacement.

AlexsanderDamaceno replied:
@cdumez I had change messageData to be a span

cdumez commented on Source/WebKit/Platform/IPC/unix/ConnectionUnix.cpp:
```diff
@@ -159,10 +159,10 @@ bool Connection::processMessage()
         return false;
 
-    uint8_t* messageData = m_readBuffer.data();
+    std::span<uint8_t> messageData(m_readBuffer.data(), m_readBuffer.size());
```
`auto messageData = m_readBuffer.mutableSpan();`

cdumez commented on Source/WebKit/Platform/IPC/unix/ConnectionUnix.cpp:
```diff
@@ -159,10 +159,10 @@ bool Connection::processMessage()
     MessageInfo messageInfo;
-    memcpy(static_cast<void*>(&messageInfo), messageData, sizeof(messageInfo));
-    messageData += sizeof(messageInfo);
+    memcpySpan(asMutableByteSpan(messageInfo), messageData.first(sizeof(messageInfo)));
```
You can use:
```
memcpySpan(asMutableByteSpan(messageInfo), consumeSpan(messageData, sizeof(messageInfo)));
```

cdumez commented on Source/WebKit/Platform/IPC/unix/ConnectionUnix.cpp:
```diff
@@ -159,10 +159,10 @@ bool Connection::processMessage()
-    memcpy(static_cast<void*>(&messageInfo), messageData, sizeof(messageInfo));
-    messageData += sizeof(messageInfo);
+    memcpySpan(asMutableByteSpan(messageInfo), messageData.first(sizeof(messageInfo)));
+    messageData = messageData.subspan(sizeof(messageInfo));
```
And then you no longer need this.

cdumez commented on Source/WebKit/Platform/IPC/unix/ConnectionUnix.cpp:
```diff
@@ -178,8 +178,8 @@ bool Connection::processMessage()
     if (attachmentCount) {
-        memcpy(static_cast<void*>(attachmentInfo.data()), messageData, sizeof(AttachmentInfo) * attachmentCount);
-        messageData += sizeof(AttachmentInfo) * attachmentCount;
+        memcpySpan(asMutableByteSpan(attachmentInfo), messageData.first(sizeof(AttachmentInfo) * attachmentCount));
```
Same thing, could use `consumeSpan()`

cdumez commented on Source/WebKit/Platform/IPC/unix/ConnectionUnix.cpp:
```diff
@@ -223,7 +223,7 @@ bool Connection::processMessage()
     ASSERT(attachments.size() == (messageInfo.isBodyOutOfLine() ? messageInfo.attachmentCount() - 1 : messageInfo.attachmentCount()));
 
-    uint8_t* messageBody = messageData;
+    uint8_t* messageBody = messageData.data();
```
Can we keep `messageBody` as a span here?

AlexsanderDamaceno replied:
Ok, changed, thanks for the suggestions.

cdumez commented on Source/WebKit/Platform/IPC/unix/ConnectionUnix.cpp:
```diff
@@ -223,11 +220,11 @@ bool Connection::processMessage()
+        messageBody = oolMessageBody->mutableSpan();
 
-    auto decoder = Decoder::create({ messageBody, messageInfo.bodySize() }, WTFMove(attachments));
+    auto decoder = Decoder::create({ messageBody.data(), messageInfo.bodySize() }, WTFMove(attachments));
```
`messageBody.first(messageInfo.bodySize())`

Creating a span from a pointer and a size is considered unsafe and should be avoided as much as possible.

cdumez reviewed (approved):
LGTM

mcatanzaro reviewed (changes requested):
.

cdumez commented on Source/WebKit/Platform/IPC/unix/ConnectionUnix.cpp:
```diff
@@ -178,9 +177,7 @@ bool Connection::processMessage()
-        memcpy(static_cast<void*>(attachmentInfo.data()), messageData, sizeof(AttachmentInfo) * attachmentCount);
-        messageData += sizeof(AttachmentInfo) * attachmentCount;
-
+        memcpySpan(asMutableByteSpan(attachmentInfo), consumeSpan(messageData, sizeof(AttachmentInfo) * attachmentCount));
```
Oh, calling `asMutableByteSpan()` is **not** the right thing to do here. I've just noticed that `attachmentInfo` is a Vector.

cdumez replied:
You likely want to do `asMutableByteSpan(attachmentInfo.mutableSpan())`

AlexsanderDamaceno replied:
Ah, OK.

cdumez reviewed (approved):
LGTM. Looks like we got there eventually. The bots seem happy now.

AlexsanderDamaceno committed:
Reduce use of memcpy in ConnectionUnix.cpp
https://bugs.webkit.org/show_bug.cgi?id=290511

Reviewed by Chris Dumez.

Reduce use of memcpy in ConnectionUnix, memcpy is unsafe and can lead to memory corruption bugs.

* Source/WebKit/Platform/IPC/unix/ConnectionUnix.cpp:
(IPC::Connection::processMessage):

Canonical link: https://commits.webkit.org/292897@main
